### PR TITLE
feat: support for vc+sd-jwt

### LINKED ISSUE #269: [Satosa] [Bug] Errore verifica corretteza in integration test
L'esito finale dell'integration test controlla gli attributi ottenuti confrontandoli con quelli attesi in questa riga di codice

https://github.com/italia/eudi-wallet-it-python/blob/1cad07f7bb3ae8d6958d3b1cb3b65bea168ff866/example/satosa/integration_test/main.py#L270-L275

questo check non è formalmente corretto. Ad esempio, se l'autenticazione non genera nessun attributo dell'utente, o li genera con nome sbagliato, il test viene comunque superato.

### PR DESCRIPTION
**Pull request by Zicchio**
This pull request closes #269 and #255.
Proposals and modifications are *very* welcomed.

The followings steps where done:
1. define a new interface for general purpose VP token (which is not intended to be a definitive and final iteration; open for review if/when we handle mdoc-cbor tokens); previous iteration "kinda" assume that the only possible VP tokens were vp-envolpes.
2. new interface is intended to be decoupled from trust-layer and direct post processing as much possible for extensibility purposes.
3. define an implementation for that interface for vc+sd-jwt tokens - **In this PR only PIDs are obtained this way**.
4. integrate the usage of the new interface in the `response_handler.py`, with some care regarding extensibility and possible modifications at the trust layers

The final iteration resulted in more modifcations that intended - special care should be taken before pulling in @peppelinux .

Note that in the process we "lost" VP-envolpe; which was expclicity *against* the intent of this pull request. This is because the envolpe hypotesys was spread around in multiple methods and I found it hard to salvage (I don't deny that it might be a skill issue). Hopelly, the new interface should promote a new implementation in short time.

### CODE REVIEW

File: pyeudiw/openid4vp/vp_sd_jwt_kb.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+from pyeudiw.jwt.utils import unsafe_parse_jws
+from pyeudiw.jwt.schemas.jwt import UnverfiedJwt
+from pyeudiw.openid4vp.exceptions import InvalidVPKeyBinding, InvalidVPSignature, KIDNotFound, VPSchemaException
+from pyeudiw.openid4vp.verifier import VpVerifier
+from pyeudiw.sd_jwt.schema import KeyBindingJwtHeader, KeyBindingJwtPayload, VcSdJwtHeaderSchema, VcSdJwtPayloadSchema, is_sd_jwt_kb_format
+from pyeudiw.tools.utils import iat_now
+
+
+_SD_JWT_DELIMITER = '~'
+
```
Comment (peppelinux) [inline pyeudiw/openid4vp/vp_sd_jwt_kb.py:21]:
```suggestion
```

File: pyeudiw/openid4vp/vp_sd_jwt_kb.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+@dataclass
+class VerifierChallenge:
+    aud: str
+    nonce: str
+
+
+class VpVcSdJwtKbVerifier(VpVerifier):
+    _DEFAULT_DISCLOSABLE_CLAIMS = (
```
Comment (peppelinux) [inline pyeudiw/openid4vp/vp_sd_jwt_kb.py:32]: please move this in the settings

File: pyeudiw/openid4vp/vp_sd_jwt_kb.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        return filtered_claims_result
+
+    def __str__(self) -> str:
+        return "VpVcSdJwtKb(" \
+            f"sdjwt={self.sdjwtkb}" \
+            ")"
+
+
+def _verify_jws_with_key(issuer_jwt: str, issuer_key: JWK):
+    verifier = JWSHelper(issuer_key)
```
[the remark this answers is not in the document]
Comment (peppelinux) [reply pyeudiw/openid4vp/vp_sd_jwt_kb.py:144]: please protect it with a specialized try/except

File: pyeudiw/satosa/default/response_handler.py [hunk 4]
```diff
@@ -62,131 +68,190 @@ def _extract_all_user_attributes(self, attributes_by_issuers: dict) -> dict:
...
-        try:
-            vpt.load_nonce(stored_session['nonce'])
-            vps: list[Vp] = vpt.get_presentation_vps()
-            vpt.validate()
+        return context.request
 
-        except VPNotFound as e:
-            _msg = "Error while retrieving VP. Payload 'vp_token' is empty or has an unexpected value."
-            return self._handle_400(context, _msg, e)
+    def _detect_typ_iss_vptoken(self, vp_token: str) -> tuple[str, str]:
```
[the remark this answers is not in the document]
Comment (peppelinux) [reply pyeudiw/satosa/default/response_handler.py:95]: here we should pass the presentation_submission to guide the _detect_typ_iss_vptoken in using the right class for decoding header/payload

**Review by peppelinux — CHANGES_REQUESTED** [review]
Good achievements, anyway we have decided to continue in this PR these two core components:

- trust layer abstraction, according to: https://github.com/italia/eudi-wallet-it-python/issues/267
- credential format abstraction, following the comments within this PR

**Comment by Zicchio** [comment]
dev note: I rebased the branch based on the changes of PR #270 and PR #272 and handles some conflicts (mostly in the ingration test hanges with PR #272).
no further changes are introduced

File: pyeudiw/openid4vp/interface.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        raise NotImplementedError
+
+    def is_active(self) -> bool:
+        return (not self.is_expired()) and (not self.is_revoked())
+
+    def verify_signature(self, public_key: JWK) -> None:
+        """
+        :raises [InvalidSignatureException]:
+        """
+        raise NotImplementedError
```
Comment (peppelinux) [inline pyeudiw/openid4vp/interface.py:46]:
```suggestion
        raise NotImplementedError

    # TODO: VP proof of possession verification method should be implemented
```

File: pyeudiw/openid4vp/vp_sd_jwt_vc.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        return iss
+
+    def get_credentials(self) -> dict:
+        return self.sdjwt.get_disclosed_claims()
+
+    def get_signing_key(self) -> JWK | KeyIdentifier_T:
+        return extract_key_identifier(self.sdjwt.issuer_jwt.header)
+
+    def is_revoked(self) -> bool:
+        return False
```
Comment (peppelinux) [inline pyeudiw/openid4vp/vp_sd_jwt_vc.py:34]:
```suggestion
        # TODO: implement revocation check
        return False
```

File: pyeudiw/satosa/default/response_handler.py [hunk 7]
```diff
@@ -62,131 +71,144 @@ def _extract_all_user_attributes(self, attributes_by_issuers: dict) -> dict:
...
-        if context.request_method.lower() != 'post':
-            return self._handle_400(context, "HTTP Method not supported")
-        _endpoint = f'{self.server_url}{context.request_uri}'
+        :return: a dictionary containing the request data
+        :rtype: dict
+
+        :raises BadRequestError: when request paramets are in a not processable state; the expected handling is returning 400
+        """
+        if (http_method := context.request_method.lower()) != ResponseHandler._SUPPORTED_RESPONSE_METHOD:
+            raise BadRequestError(f"HTTP method [{http_method}] not supported")
```
Comment (peppelinux) [inline pyeudiw/satosa/default/response_handler.py:86]:
```suggestion
        # TODO: post is required for IT-Wallet, if not post get must work anyway
        # if (http_method := context.request_method.lower()) != ResponseHandler._SUPPORTED_RESPONSE_METHOD:
        #    raise BadRequestError(f"HTTP method [{http_method}] not supported")
```

File: pyeudiw/satosa/default/response_handler.py [hunk 8]
```diff
@@ -62,131 +71,144 @@ def _extract_all_user_attributes(self, attributes_by_issuers: dict) -> dict:
...
+
+        # the flow below is a simplified algorithm of authentication response processing, where:
+        # (1) we don't check that presentation submission matches definition (yet)
+        # (2) we don't check that vp tokens are aligned with information declared in the presentation submission
+        # (3) we use all disclosed claims in vp tokens to build the user identity
+        attributes_by_issuer: dict[str, dict[str, Any]] = {}
+        credential_issuers: list[str] = []
+        encoded_vps: list[str] = [authz_payload.vp_token] if isinstance(authz_payload.vp_token, str) else authz_payload.vp_token
+        for vp_token in encoded_vps:
+            # verify vp token and extract user information
```
Comment (peppelinux) [inline pyeudiw/satosa/default/response_handler.py:181]:
```suggestion
            # verify vp token and extract user information
            # TODO: specialized try/except for each call, from line 182 to line 187
```

File: pyeudiw/tests/sd_jwt/test_sdjwt.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    "InNoYS0yNTYiLCAiY25mIjogeyJqd2siOiB7Imt0eSI6ICJFQyIsICJjcnYiOiAiUC0y" \
+    "NTYiLCAieCI6ICJUQ0FFUjE5WnZ1M09IRjRqNFc0dmZTVm9ISVAxSUxpbERsczd2Q2VH" \
+    "ZW1jIiwgInkiOiAiWnhqaVdXYlpNUUdIVldLVlE0aGJTSWlyc1ZmdWVjQ0U2dDRqVDlG" \
+    "MkhaUSJ9fX0.ZfSxIFLHf7f84WIMqt7Fzme8-586WutjFnXH4TO5XuWG_peQ4hPsqDpi" \
+    "MBClkh2aUJdl83bwyyOriqvdFra-bg"
+
+DISCLOSURES = [
+    "WyJlbHVWNU9nM2dTTklJOEVZbnN4QV9BIiwgImZhbWlseV9uYW1lIiwgIkRvZSJd",
+    "WyJBSngtMDk1VlBycFR0TjRRTU9xUk9BIiwgImFkZHJlc3MiLCB7InN0cmVldF9hZGRy" +
+        "ZXNzIjogIjEyMyBNYWluIFN0IiwgImxvY2FsaXR5IjogIkFueXRvd24iLCAicmVnaW9u" +
```
Comment (peppelinux) [inline pyeudiw/tests/sd_jwt/test_sdjwt.py:105]:
```suggestion
    "WyJBSngtMDk1VlBycFR0TjRRTU9xUk9BIiwgImFkZHJlc3MiLCB7InN0cmVldF9hZGRy", \
    "ZXNzIjogIjEyMyBNYWluIFN0IiwgImxvY2FsaXR5IjogIkFueXRvd24iLCAicmVnaW9u", \
```

File: pyeudiw/tests/sd_jwt/test_sdjwt.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        "IjogIkFueXN0YXRlIiwgImNvdW50cnkiOiAiVVMifV0",
```
Comment (peppelinux) [inline pyeudiw/tests/sd_jwt/test_sdjwt.py:106]:
```suggestion
        "IjogIkFueXN0YXRlIiwgImNvdW50cnkiOiAiVVMifV0",
```

File: pyeudiw/tests/settings.py [hunk 11]
```diff
@@ -341,7 +341,309 @@
...
+        "error_template": "error.html",
+        "error_url": "https://localhost:9999/error_page.html"
+    },
+    "endpoints": {
+        "entity_configuration": "/.well-known/openid-federation",
+        "pre_request": "/pre-request",
+        "response": "/response-uri",
+        "request": "/request-uri",
+        "status": "/status-uri",
+        "get_response": "/get-response",
```
Comment (peppelinux) [inline pyeudiw/tests/settings.py:378]:
```suggestion
        "get_response": "/get-response"
```

File: pyeudiw/tests/settings.py [hunk 12]
```diff
@@ -341,7 +341,309 @@
...
+    },
+    "jwt": {
+        "default_sig_alg": "ES256",
+        "default_exp": 6
+    },
+    "authorization": {
+        "url_scheme": "haip",  # haip://
+        "scopes": ["pid-sd-jwt:unique_id+given_name+family_name"],
+        "default_acr_value": "https://www.spid.gov.it/SpidL2",
+        "expiration_time": 5,  # minutes
```
Comment (peppelinux) [inline pyeudiw/tests/settings.py:397]:
```suggestion
        "expiration_time": 5  # minutes
```

File: pyeudiw/tests/trust/test_dynamic.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        pass
+
+    def get_public_keys(self, issuer: str) -> list[dict]:
+        return [
+            MockTrustEvaluator.mock_jwk
+        ]
+
+    def get_metadata(self, issuer: str) -> dict:
+        return {
+            "jsono_key": "json_value"
```
Comment (peppelinux) [inline pyeudiw/tests/trust/test_dynamic.py:29]:
```suggestion
            "json_key": "json_value"
```

File: pyeudiw/tests/trust/test_dynamic.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            "config": {
+                "jwk_endpoint": "/.well-known/jwt-vc-issuer",
+                "httpc_params": {
+                    "connection": {
+                        "ssl": True
+                    },
+                    "session": {
+                        "timeout": 6
+                    }
+                }
```
Comment (peppelinux) [inline pyeudiw/tests/trust/test_dynamic.py:58]: please use `__init__.DEFAULT_DIRECT_TRUST_PARAMS`

**Review by peppelinux — APPROVED** [review]

**Pull request merged**